doc/source/dev/api-dev/array_api.rst

```diff
@@ -0,0 +1,253 @@
+
+    >>> # The above uses numpy arrays; trying to use np.matrix instances or object
+    >>> # arrays instead will yield an exception with `SCIPY_ARRAY_API=1`:
+    >>> vq(np.asmatrix(features), code_book)
+    ...
+    TypeError: 'numpy.matrix' are not supported
+
+    >>> vq(np.ma.asarray(features), code_book)
+    ...
+    TypeError: 'numpy.ma.MaskedArray' are not supported
+
+    >>> vq(features.astype(np.object_), code_book)
+    ...
+    TypeError: object arrays are not supported
+
+
+Currently supported functionality
+`````````````````````````````````
+
+The following modules provide Array API standard support when the environment
+variable is set:
+
+- ``scipy.cluster.hierarchy``
+- ``scipy.cluster.vq``
+
+
+Implementation notes
+--------------------
+
+A key part of the support for the array API standard and specific compatibility
+functions for Numpy, CuPy and PyTorch is provided through
+`array-api-compat <https://github.com/data-apis/array-api-compat>`_.
+This package is included in the SciPy code base via a git submodule (under
+``scipy/_lib``), so no new dependencies are introduced.
+
+``array-api_compat`` provides generic utility functions and adds aliases such
+as ``xp.concat`` (which, for numpy, maps to ``np.concatenate``). This allows
+using a uniform API across NumPy, PyTorch and CuPy (as of right now; support
+for other libraries like JAX is expected to be added in the future).
+
+When the environment variable isn't set and hence Array API support in SciPy is
+disabled, we still use the "augmented" version of the NumPy namespace, which is
+``array_api_compat.numpy`` (this should not change behavior of SciPy functions).
+When support is enabled, depending on the type of arrays, ``xp`` will return the
+standard-compatible namespace matching the input array type to a function (e.g.,
+if the input to ``cluster.vq.kmeans`` is a PyTorch array, then ``xp`` is
+``array_api_compat.torch``).
+
+
+Adding Array API support to a SciPy function
+--------------------------------------------
+
+As much as possible, new code added to SciPy should try to follow as closely as
+possible the Array API standard (these functions typically are best-practice
+idioms for NumPy usage as well). By following the standard, effectively adding
+support for Array API is typically straightforward, and we ideally don't need
+to maintain any customization.
+
+Two helper functions are available:
+
+* ``array_namespace``: detect the namespace based on input arrays and do some
+  input validation (like refusing to work with masked arrays, please see the
+  `RFC`_.)
+* ``as_xparray``: a drop-in replacement for ``np.asarray`` with additional
+  features like ``copy, check_finite``. As stated above, try to limit the use
+  of non standard features. In the end we would want to upstream our needs to
+  the compatibility library.
+
+To add support to a Python-level SciPy function, what you have to change is:
```

Comment:
```suggestion
To add support to a Python-level SciPy function, i.e. no compiled code involved, what you have to change is:
```

Reply:
This seems incorrect, given point (3) below. It's any function that has a Python `def ...` entry point. I'll rephrase.